![](images/vaccine-6116391_1920_pixabay.jpg)

***
***

# **COVID-19 Vaccination Progress - EDA**
## *Data Cleaning and Feature Engineering*

***
***

## Table of Contents

| 1. INTRODUCTION | 2. IMPORTS AND DATA PREPARATION | 3. IN-DEPTH DATA ANALYSIS AND VISUALISATION | 4. CONCLUSION |
|----|----|----|----|
| *1.1 COVID-19 pandemic* | *2.1 Importing libraries and data* | ADD LINK TO VIZ NOTEBOOK | ADD LINK TO VIZ NOTEBOOK |
| *1.2 COVID-19 vaccines* | *2.2 Preliminary data exploration and subsequent cleaning* | | |
| *1.3 Key objectives* |  *2.3 Feature engineering to prepare for in-depth data analysis* | | |
| *1.4 Dataset overview* | | | |

***
***

## 1. INTRODUCTION

***

### 1.1 COVID-19 pandemic

The coronavirus disease (COVID-19) is an infectious disease which is caused by a new coronavirus and was first detected in Q4 2019. As the disease spread around the world it was declared a pandemic early in 2020. Now, a year later (March 2021), the effects of the global pandemic continues to affect and change the lives of each and everyone across the world. Whereas most people who contract the disease will experience mild to moderate respiratory illness and recover without requiring special treatment, especially the elderly, and those with underlying medical problems (e.g. cardiovascular disease, diabetes, chronic respiratory disease, cancer) are more likely to develop very serious conditions. The way forward in fighting the disease it to vaccinated the world population. The progress of this process over the last 3 months is what we will investigatee in this EDA.

A complete overview of the COVID-19 pandemic by the [WHO](https://www.who.int/health-topics/coronavirus#tab=tab_1).

***

### 1.2 COVID-19 vaccines

The speed at which the first COVID-19 vaccines were developed as a response to the pandemic was extraordinary. Overall it took less than a year to successfully develop and approve several vaccines, a timeframe significantly shorter than usually observed for the development of new vaccines. It is important that development of further vaccines continues, as it does, as a very large share of the world population will need to receive a COVID-19 vaccine in order to become immune to the virus, so that we can successfully combat the pandemic. In addition, the appearance of more and more mutated versions of the SARS-CoV-2 virus calls for the development of additional vaccines to offer a wide bandwidth of starting points to also sucessfully combat these mutants. As such, one of the key challenges is to vaccinate people around the world. ['OurWorldinData'](https://ourworldindata.org/covid-vaccinations) create and make available a daily updated dataset on international COVID-19 vaccinations to track these efforts.

There are different types of vaccines developed to target SARS-CoV-2, RNA based vaccines and cell based vaccines (recombinant and inactivated) and protein subunit based vaccines. An overview of the different types can be found [here](https://covid19.trackvaccines.org/how-do-vaccines-work/#types).

mRNA VACCINES: mRNA vaccines contain mRNA with instructions for making the spike protein. This protein is expressed the surface of the SARS-CoV-2 virus and required to enter cells. Upon vaccination the spike protein is produced and will be recognised by the immune system as foreign. This results in the production of antibodies and activated T cells (white blood cells) which will attack it. This information is stored by the immune system so that if it encounters the SARS-CoV-2 virus again it can act fast to eliminate it.

RECOMBINANT VACCINES: Recombinant vaccines are made up of another virus (which cannot reproduce and cause disease) that has been modified to contain the gene for making the SARS-CoV-2 spike protein. This protein is expressed the surface of the SARS-CoV-2 virus and required to enter cells. Upon vaccination the spike protein is produced and will be recognised by the immune system as foreign. This results in the production of antibodies and activated T cells (white blood cells) which will attack it. This information is stored by the immune system so that if it encounters the SARS-CoV-2 virus again it can act fast to eliminate it.

Several institutions maintain websites on which they list COVID-19 candidate vaccines that are currently being developed. The [WHO Vaccine Tracker](https://www.who.int/publications/m/item/draft-landscape-of-covid-19-candidate-vaccines) is one of them and provides a concise overview. Additional information on COVID-19 vaccines by the WHO can be found [here](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/covid-19-vaccines).


At this point an overview of the approved vaccines will be given for selected countries only (USA, EU, China, India, Israel, Russia) [status: 31st March 2021].

General overview of the current (31st March 2021) top 4 vaccines with respect to the number of countries in which they received authorisation:

| VACCINE | Approved in | type of vaccine |
|----|----|----|
| Oxford/AstraZeneca | 82 countries | RECOMBINANT (non-replicating viral vector) |
| Pfizer/BioNTech | 79 countries | messenger RNA (mRNA) (nucleoside-modified) |
| Johnson & Johnson (Janssen) | 37 countries | RECOMBINANT (non-replicating viral vector) |
| Covishield (Serum Inst. of India) | 31 countries | RECOMBINANT (non-replicating viral vector) |


#### USA

+ [FDA vaccine overview](https://www.fda.gov/emergency-preparedness-and-response/coronavirus-disease-2019-covid-19/covid-19-vaccines)

| VACCINE | status | type of vaccine | administration |
|----|----|----|----|
| Pfizer/BioNTech | authorised for emergency use (11/12/2020) | RECOMBINANT (non-replicating viral vector) | 2 intramuscular injections, 4-12 weeks apart |
| Moderna | authorised for emergency use (18/12/2020) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 28 days apart |
| Johnson & Johnson (Janssen) | authorised for emergency use (27/02/2021) | RECOMBINANT (non-replicating viral vector) |


#### European Union

NOTE: The following summary reflects the approved vaccines in the EU, not Europe. In Switzerland the Oxford/AstraZeneca vaccine is not approved, and in the UK not Johnson & Johnson is not approved.

+ [EMA vaccine overview](https://www.ema.europa.eu/en/human-regulatory/overview/public-health-threats/coronavirus-disease-covid-19/treatments-vaccines/vaccines-covid-19/covid-19-vaccines-key-facts)

+ [Authorised vaccine overview - European medicines agency - EMA](https://www.ema.europa.eu/en/human-regulatory/overview/public-health-threats/coronavirus-disease-covid-19/treatments-vaccines/vaccines-covid-19/covid-19-vaccines-authorised)

| VACCINE | status | type of vaccine | administration |
|----|----|----|----|
| Pfizer/BioNTech (Corminaty) | conditional marketing authorisation granted (21/12/2020) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 3 weeks apart |
| Johnson & Johnson (Janssen) | conditional marketing authorisation granted (11/03/2021) | RECOMBINANT (non-replicating viral vector) |
| Moderna | conditional marketing authorisation granted (06/01/2021) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 28 days apart |
| AstraZeneca (Vaxzevria) | conditional marketing authorisation granted (29/01/2021) | RECOMBINANT (non-replicating viral vector) | 2 intramuscular injections, 4-12 weeks apart |


#### China, India, Israel and Russia

| VACCINE | type of vaccine | country |
|----|----|----|
| Sinopharm Wuhan | INACTIVATED virus | China |
| Sinopharm Beijing | INACTIVATED virus | China |
| Covaxin (Bharat Biotech) | INACTIVATED virus | China, India |
| CanSino | RECOMBINANT (non-replicating viral vector) | China |
| Anhui Zhifei Longcom (RBD-Dimer) | Protein subunit | China |
| Oxford/AstraZeneca | RECOMBINANT (non-replicating viral vector) | India |
| Covishield (Serum Inst. of India) | RECOMBINANT (non-replicating viral vector) | India |
| Pfizer/BioNTech | messenger RNA (mRNA) (nucleoside-modified) | Israel |
| Moderna | messenger RNA (mRNA) (nucleoside-modified) | Israel |
| Sputnik V | RECOMBINANT (non-replicating viral vector) | Russia |
| EpiVacCorona | Protein subunit | Russia |

An overview of clinical trials and vaccine authorisations by country can be found [here](https://covid19.trackvaccines.org/trials-vaccines-by-country/).

An overview of the different types of vaccines can be found [here](https://www.who.int/news-room/feature-stories/detail/the-race-for-a-covid-19-vaccine-explained).

In summary it can be noted that out of the total 12 vaccines listed above to date only 10 appear in the dataset, i.e. Covaxin, EpiVacCorona, Johnson & Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm Beijing, Sinopharm Wuhan, Sinovac, Sputnik V. Two chinese vaccines CanSino and Anhui Zhifei Longcom are not currently covered by the dataset. It can also be noted that some vaccines are approved accross numerous countries, whereas others are predominantly appoved in only a limited number of countries. The above overview of vaccines is not exhaustive and there may be some approved vaccines that are not included. And, of course, the approval of new vaccines is a dynamic progress, so the overview only reflects the status as of 31st March 2021.

IMPORTANT NOTE! Different vaccines are approved for use in different countries and thus influence which vaccines and how many are available in each country. In addition the available is strongly influenced by the production capacity of each vaccine.

***

### 1.3 Key objectives

***

### 1.4 Dataset overview

The dataset used in this capstone project is COVID-19 World Vaccination Progress (version 81) created by Gabriel Preda and [available on Kaggle](https://www.kaggle.com/gpreda/covid-world-vaccination-progress). It is a compilation of COVID-19 vaccination data provided by [OurWordinData](https://ourworldindata.org/covid-vaccinations). The 'OurWorldinData' website contains a plethory of information regarding the global vaccination progress, including visualisations of the progress as well as an overview of approved vaccines worldwide. 

**The dataset contains 8079 observations for 15 features.**

| FEATURE | description | comment |
|----|----|----|
| **country** | 147 countries | 4 duplicates, UN recognises 197 countries (193 are members) |
| **iso_code** | country code by ISO norm | -|
| **date** | data entry (13th December 2020 to 24th March 2021) | - |
| **vaccines** | name/company | - |
| **daily_vaccinations** | no. of vaccination for that date/country | some days only daily vaccinations, some days only the (cummulative) total |
| **daily_vaccinations_raw** | daily vaccinations as CHANGE COMPARED TO PREVIOUS DAY (cross-check purposes) | calculation: total_vaccinations (day) - total_vaccination (previous day) |
| **daily_vaccinations_per_million** | ratio (in ppm) between vaccination number and total population for the current date in the country | - |
| **total_vaccinations** | absolute number of immunisations (jabs) in the country | some days only daily vaccinations, some days only the (cummulative) total |
| **total_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population up to the date in the country | - |
| **people_vaccinated** | total no. of people with immunisation (partial and full) | - |
| **people_vaccinated_per_hundred** | ratio (in percent) between population immunized (partial and full) and total population | - |
| **people_fully_vaccinated** | total no. of people with full immunisation (typically 2) | - |
| **people_fully_vaccinated_per_hundred** | ratio (in percent) between population fully immunized and total population | - |
| **source_name** | national authority, international organization, local organization etc. | - |
| **source_website** | website link to source | - |

***
***

## 2. IMPORTS AND DATA PREPARATION

***

### 2.1 Importing libraries and data

+ Import numpy, pandas and visualisation libraries (matplolib, seaborn, plotly).
+ Import Kaggle COVID vaccination progress dataset.

In [1]:
#Import basic libraries
import numpy as np
import pandas as pd
from datetime import datetime

#from pandas_profiling import ProfileReport

#Suppress pandas future warnings
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
#Import Kaggle dataset country_vaccinations on COVID-19 vaccination progress
df = pd.read_csv('data/country_vaccinations.csv')

***

### 2.2 Preliminary data exploration and subsequent cleaning

In this section the basic characteristics of the data were explored, starting with the number of observations and features, as well as the types of variables, and any missing values and/or duplicates. Detailed attention to missing values from a time-series perspective was given. Furthermore, data aggregations was carried out to prepare the data for visualisation.

Get an initial feel for the data by looking at the first 3 rows of the dataframe:

In [3]:
#Take a first glance at the data (first 3 entries)
df.head(3)

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            35.0   
2                                  NaN                            35.0   

             vaccines                source_name  \
0  Oxford/AstraZeneca  Government of Afghanistan   
1  Oxford/AstraZeneca  Government of Afghanistan   
2  Oxford/AstraZeneca  Government of Afghanistan   

                                      source_website  
0  http://www.xinhuanet.com/english/asiapacific/2...  
1  http://www.xinhuanet.com/english/asiapacific/2...  
2  http://www.xinhuanet.com/english/asiapacific/2...

Get an initial feel for the data by looking at the last 3 rows of the dataframe:

In [4]:
#Take a first glance at the data (last 3 entries)
df.tail(3)

country iso_code        date  total_vaccinations  people_vaccinated  \
8076  Zimbabwe      ZWE  2021-03-22             43294.0            43294.0   
8077  Zimbabwe      ZWE  2021-03-23             44135.0            44135.0   
8078  Zimbabwe      ZWE  2021-03-24             49404.0            49404.0   

      people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
8076                      NaN                   565.0               805.0   
8077                      NaN                   841.0               655.0   
8078                      NaN                  5269.0              1400.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
8076                            0.29                           0.29   
8077                            0.30                           0.30   
8078                            0.33                           0.33   

      people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
8076                                  NaN                            54.0   
8077                                  NaN                            44.0   
8078                                  NaN                            94.0   

               vaccines         source_name  \
8076  Sinopharm/Beijing  Ministry of Health   
8077  Sinopharm/Beijing  Ministry of Health   
8078  Sinopharm/Beijing  Ministry of Health   

                                         source_website  
8076  https://twitter.com/MoHCCZim/status/1374795915...  
8077  https://twitter.com/MoHCCZim/status/1374795915...  
8078  https://twitter.com/MoHCCZim/status/1374795915...

In [5]:
#Number of columns and rows
shape1 = df.shape
print(f'NOTE: In total we have {shape1[1]} columns and {shape1[0]} rows in our COVID vaccination dataset.')

NOTE: In total we have 15 columns and 8079 rows in our COVID vaccination dataset.


In [6]:
#Generate count statistics of duplicate entries
if len(df[df.duplicated()]) > 0:
    print("NOTE: No. of duplicated entries in dataframe: ", len(df[df.duplicated()]))
    print(df[df.duplicated(keep=False)].sort_values(by=list(df.columns)).head())
else:
    print("NOTE: No duplicated entries found in COVID vaccination progress dataset.")

NOTE: No duplicated entries found in COVID vaccination progress dataset.


In [7]:
#Information about datatypes per column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8079 entries, 0 to 8078
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              8079 non-null   object 
 1   iso_code                             8079 non-null   object 
 2   date                                 8079 non-null   object 
 3   total_vaccinations                   4984 non-null   float64
 4   people_vaccinated                    4423 non-null   float64
 5   people_fully_vaccinated              3021 non-null   float64
 6   daily_vaccinations_raw               4192 non-null   float64
 7   daily_vaccinations                   7891 non-null   float64
 8   total_vaccinations_per_hundred       4984 non-null   float64
 9   people_vaccinated_per_hundred        4423 non-null   float64
 10  people_fully_vaccinated_per_hundred  3021 non-null   float64
 11  daily_vaccinations_per_million

+ NOTE: 9 out of 15 columns contain null values and need to be dealt with.
+ Therefore, let's take a look at how many empties we have to deal with.
+ And, let's check if there are any empty strings or lists in the 'object' columns. 

In [8]:
df.isnull().sum(axis = 0)

country                                   0
iso_code                                  0
date                                      0
total_vaccinations                     3095
people_vaccinated                      3656
people_fully_vaccinated                5058
daily_vaccinations_raw                 3887
daily_vaccinations                      188
total_vaccinations_per_hundred         3095
people_vaccinated_per_hundred          3656
people_fully_vaccinated_per_hundred    5058
daily_vaccinations_per_million          188
vaccines                                  0
source_name                               0
source_website                            0
dtype: int64

In [9]:
#Check for empty strings/lists in object columns
object_lst = ['country', 'iso_code', 'date', 'vaccines', 'source_name', 'source_website']
lst_empties = ['', []]
check_str_lst = df[object_lst].isin(lst_empties).value_counts()
print(f'The check for empty strings/lists in the object columns returns: \n \n {check_str_lst}')

The check for empty strings/lists in the object columns returns: 
 
 country  iso_code  date   vaccines  source_name  source_website
False    False     False  False     False        False             8079
dtype: int64


In [10]:
#Basic statistics (1)
df.describe().transpose().round()

count       mean        std  min  \
total_vaccinations                   4984.0  2597915.0  9383280.0  0.0   
people_vaccinated                    4423.0  2073454.0  6824277.0  0.0   
people_fully_vaccinated              3021.0   893378.0  3595095.0  1.0   
daily_vaccinations_raw               4192.0    92034.0   289560.0  0.0   
daily_vaccinations                   7891.0    60968.0   213914.0  1.0   
total_vaccinations_per_hundred       4984.0       10.0       18.0  0.0   
people_vaccinated_per_hundred        4423.0        8.0       12.0  0.0   
people_fully_vaccinated_per_hundred  3021.0        4.0        8.0  0.0   
daily_vaccinations_per_million       7891.0     2699.0     4163.0  0.0   

                                         25%       50%        75%          max  
total_vaccinations                   40452.0  276962.0  1303350.0  130473853.0  
people_vaccinated                    36780.0  249555.0  1007708.0   85472166.0  
people_fully_vaccinated              19019.0  111716.0   488937.0   46365515.0  
daily_vaccinations_raw                2738.0   13838.0    56702.0    4575496.0  
daily_vaccinations                    1026.0    6161.0    27794.0    2541597.0  
total_vaccinations_per_hundred           1.0       4.0       11.0        162.0  
people_vaccinated_per_hundred            1.0       3.0        9.0         90.0  
people_fully_vaccinated_per_hundred      0.0       2.0        3.0         72.0  
daily_vaccinations_per_million         363.0    1305.0     3211.0      54264.0

In [11]:
#Basic statistics (2)
object_lst = ['country', 'iso_code', 'date', 'vaccines', 'source_name', 'source_website']
df[object_lst].astype('object').describe().transpose()

count unique  \
country         8079    151   
iso_code        8079    151   
date            8079    102   
vaccines        8079     26   
source_name     8079     92   
source_website  8079    147   

                                                              top  freq  
country                                                  Scotland   101  
iso_code                                                 OWID_SCT   101  
date                                                   2021-03-10   133  
vaccines             Moderna, Oxford/AstraZeneca, Pfizer/BioNTech  1984  
source_name                                    Ministry of Health  2569  
source_website  https://coronavirus.data.gov.uk/details/health...   505

In [12]:
#Number of countries included in dataset:
print(F'{df.country.nunique()} countries are included in the Covid-19 vaccination progress dataset.')

151 countries are included in the Covid-19 vaccination progress dataset.


In [13]:
df.country.unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Anguilla', 'Antigua and Barbuda', 'Argentina', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Bermuda', 'Bolivia',
       'Brazil', 'Bulgaria', 'Cambodia', 'Canada', 'Cayman Islands',
       'Chile', 'China', 'Colombia', 'Costa Rica', "Cote d'Ivoire",
       'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Dominica',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'England',
       'Equatorial Guinea', 'Estonia', 'Faeroe Islands',
       'Falkland Islands', 'Finland', 'France', 'Germany', 'Ghana',
       'Gibraltar', 'Greece', 'Greenland', 'Grenada', 'Guatemala',
       'Guernsey', 'Guinea', 'Guyana', 'Honduras', 'Hong Kong', 'Hungary',
       'Iceland', 'India', 'Indonesia', 'Iran', 'Ireland', 'Isle of Man',
       'Israel', 'Italy', 'Jamaica', 'Japan', 'Jersey', 'Jordan',
       'Kazakhstan', 'Kenya', 'Kuwait', 'Laos

In [14]:
#TODO test this out

#def flat_unique_elements(array_of_array):
#    vaccine_list = []
#    for childArray in array_of_array:
#        for child in childArray.split(","):
#            if child.strip() not in vaccine_list:
#                vaccine_list.append(child.strip())
#    return vaccine_list

#vaccines = flat_unique_elements(df['vaccines'].unique())

#print("list of all vaccines  -\n", vaccines)


#Number of days of vaccination campaingn
#country_days_of_vaccination =df.country.where(df.people_vaccinated>0).value_counts()
#country_days_of_vaccination.plot()

#Plotly
#fig = px.line(vaccin, x="date", y='daily_vaccinations', color='country')
#fig.show()

#Vaccines used in different countries
#fig = px.choropleth(
#         country_vaccines, # Passing the dataframe 
#         locations='country', # Select the column with the name of the countries
#         color='vaccines', # Passing a column with values
#         locationmode='country names' # We pass the parameter of determining the country on the map (by name)
#)

#fig.update_layout(
    # Set the name of the map
#    title_text='<b>Vaccines used in different countries</b>',
    # Hide the legend
#    showlegend=False,
    # Determine the map display settings (remove the frame, etc.)
#    geo=dict(
#       showframe=False,
#       showcoastlines=False,
#       projection_type='equirectangular'
#    ),
       # Setting parameters for the text
#       font=dict(
#           family='Arials',
#           size=13, 
#           color='black'
#    )
#)
#fig.show()

The 'country' column appears to contain duplicates (England, United Kingdom, Northern Ireland, Wales, Scotland).

***
**SUMMARY - CLEANING TASKS** 

- [ ] Drop 'daily_vaccinations_raw' column as only for cross-check purposes, i.e. information contained in total_vaccinations column

- [ ] Drop 'source_name', 'source_website' as not required for data analysis (Note: Keep ISO for worldmap plots)

- [ ] Change object to datetime for column date to check first/last entry date (Note: Conversion gets lost upon save as csv)

- [ ] Investigate and tidy-up 'vaccines' column

- [ ] Check 'country' column in detail as it contains duplicates (England, United Kingdom, Northern Ireland, Wales, Scotland)

- [ ] Check columns with missing values in more detail

***

In [15]:
#Copy dataset before cleaning
df_1 = df.copy()

In [16]:
#Drop columns daily_vaccinations_raw, iso_code, source_name, source_website
lst = ['daily_vaccinations_raw', 'source_name', 'source_website']
df_1.drop(lst, 1, inplace=True)

- [x] Dropped 'daily_vaccinations_raw' column as only for cross-check purposes

- [x] Dropped 'source_name' and 'source_website' as not required for data analysis

In [17]:
#Change type for date to datetime
df_1['date'] =  pd.to_datetime(df_1['date'])

- [x] Changed object to datetime for column date to check first/last entry date

In [18]:
#Check for first/last entry date
d_min = df_1.date.min()
d_max = df_1.date.max()
print(f'The dataset covers vaccinations between {d_min} and {d_max}.')

The dataset covers vaccinations between 2020-12-13 00:00:00 and 2021-03-24 00:00:00.


In [19]:
#Check time delta
t_delta = df_1.date.max() - df_1.date.min()
print(f'The dataset covers {t_delta}.')

The dataset covers 101 days 00:00:00.


In [20]:
#Confirm changes
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8079 entries, 0 to 8078
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              8079 non-null   object        
 1   iso_code                             8079 non-null   object        
 2   date                                 8079 non-null   datetime64[ns]
 3   total_vaccinations                   4984 non-null   float64       
 4   people_vaccinated                    4423 non-null   float64       
 5   people_fully_vaccinated              3021 non-null   float64       
 6   daily_vaccinations                   7891 non-null   float64       
 7   total_vaccinations_per_hundred       4984 non-null   float64       
 8   people_vaccinated_per_hundred        4423 non-null   float64       
 9   people_fully_vaccinated_per_hundred  3021 non-null   float64       
 10  daily_vaccin

NEXT TASK: Detailed investigation into vaccines column. The information is stored as a list for each country (& date) containing all vaccines used up to the respective date.

In [21]:
#Take a look at how the vaccine a country uses are stored in the vaccines column
df_1['vaccines'].unique()

array(['Oxford/AstraZeneca', 'Pfizer/BioNTech', 'Sputnik V',
       'Oxford/AstraZeneca, Sinopharm/Beijing, Sputnik V',
       'Oxford/AstraZeneca, Pfizer/BioNTech',
       'Moderna, Oxford/AstraZeneca, Pfizer/BioNTech', 'Sinovac',
       'Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sputnik V',
       'Oxford/AstraZeneca, Sinovac', 'Sinopharm/Beijing',
       'Pfizer/BioNTech, Sinovac',
       'Sinopharm/Beijing, Sinopharm/Wuhan, Sinovac',
       'Moderna, Pfizer/BioNTech', 'Moderna',
       'Moderna, Oxford/AstraZeneca',
       'Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sputnik V',
       'Covaxin, Oxford/AstraZeneca',
       'Pfizer/BioNTech, Sinopharm/Beijing',
       'Sinopharm/Beijing, Sputnik V',
       'Oxford/AstraZeneca, Pfizer/BioNTech, Sputnik V',
       'Oxford/AstraZeneca, Sinopharm/Beijing', 'EpiVacCorona, Sputnik V',
       'Johnson&Johnson', 'Pfizer/BioNTech, Sputnik V',
       'Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, S

The vaccines column was converted to one column per vaccine using booleans to indicate whether a specific vaccine is used by a specific country.

In [22]:
#Convert each entry in vaccines list to an individual row for a date/country
#and pivot to give dataframe no. 2 with individual vaccines, country and date columns
df_processed = df_1.drop('vaccines', axis=1).join(df['vaccines'].str.split(', ', expand=True).stack().reset_index(level=1, drop=True).rename('vaccines'))
df_processed = df_processed[['country', 'date', 'vaccines']].pivot(index=['country', 'date'], columns='vaccines', values='vaccines').reset_index()

#Create dataframe w/o country, date and vaccine columns
lst_ = ['country', 'date', 'vaccines']
df_2 = df_1.drop(lst_, 1)

#Join the two dataframes
df_processed = df_processed.join(df_2)

In [23]:
#Convert vaccine name entries to boolean values
#Create list of vaccines
v_lst = ', '.join(df.vaccines.tolist())
v_lst = list(set(v_lst.split(', ')))

#Replace NaN values by 0
df_processed[v_lst] = df_processed[v_lst].fillna(0)

In [24]:
#Replace vaccines by 1
vaccine_lst = ['Covaxin', 'EpiVacCorona', 'Johnson&Johnson', 'Moderna', 'Oxford/AstraZeneca', 'Pfizer/BioNTech',
                'Sinopharm/Beijing', 'Sinopharm/Wuhan', 'Sinovac', 'Sputnik V']
for c in vaccine_lst:
    df_processed[c] = df_processed[c].replace([c],1)
df_processed.sample(3)

country       date  Covaxin  EpiVacCorona  Johnson&Johnson  Moderna  \
7070       Spain 2021-01-07        0             0                0        1   
2293     England 2021-01-06        0             0                0        0   
4474  Luxembourg 2021-02-02        0             0                0        1   

      Oxford/AstraZeneca  Pfizer/BioNTech  Sinopharm/Beijing  Sinopharm/Wuhan  \
7070                   1                1                  0                0   
2293                   1                1                  0                0   
4474                   1                1                  0                0   

      ...  Sputnik V  iso_code total_vaccinations  people_vaccinated  \
7070  ...          0       ESP           207323.0           207323.0   
2293  ...          0  OWID_ENG                NaN                NaN   
4474  ...          0       LUX            14252.0            11925.0   

      people_fully_vaccinated  daily_vaccinations  \
7070                      NaN             41496.0   
2293                      NaN             98451.0   
4474                   2327.0               675.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
7070                            0.44                           0.44   
2293                             NaN                            NaN   
4474                            2.28                           1.91   

      people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
7070                                  NaN                           888.0  
2293                                  NaN                          1749.0  
4474                                 0.37                          1078.0  

[3 rows x 21 columns]

In [25]:
#Convert column header to pythonic standard (lowercase, no special characters)
df_processed.columns = df_processed.columns.str.lower().str.replace('&', '_').str.replace('/', '_').str.replace(' ', '_')

#Display sample from current dataframe
df_processed.sample(2)

country       date  covaxin  epivaccorona  johnson_johnson  moderna  \
4728   Mauritius 2021-02-08        0             0                0        0   
683   Bangladesh 2021-03-14        0             0                0        0   

      oxford_astrazeneca  pfizer_biontech  sinopharm_beijing  sinopharm_wuhan  \
4728                   1                0                  0                0   
683                    1                0                  0                0   

      ...  sputnik_v  iso_code total_vaccinations  people_vaccinated  \
4728  ...          0       MUS                NaN                NaN   
683   ...          0       BGD          4398094.0          4398094.0   

      people_fully_vaccinated  daily_vaccinations  \
4728                      NaN               227.0   
683                       NaN            102277.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
4728                             NaN                            NaN   
683                             2.67                           2.67   

      people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
4728                                  NaN                           178.0  
683                                   NaN                           621.0  

[2 rows x 21 columns]

- [X] Investigated 'vaccines' column and generated one column per vaccine indicating whether or not a country uses a specific vaccine.

NEXT TASK: Detailed investigation and cleaning of 'country' column.

In [26]:
#Spotcheck that the numbers for UK are equals to the sum for England, Scotland, Wales and northern Ireland.
lst2 = ['England', 'Scotland', 'Wales', 'Northern Ireland']
check1 = df.loc[df['country'].isin(lst2)].loc[df['date'] == '2021-03-15'].loc[:, 'daily_vaccinations'].sum()
check2 = df.loc[df['country'] == 'United Kingdom'].loc[df['date'] == '2021-03-15'].loc[:, 'daily_vaccinations'].values[0]
print(f'SPOTCHECK: The sum of daily vaccinations in England, Scotland, Wales and Nothern Ireland on 15th March 2021 was: {check1}')
print(f'SPOTCHECK: The total daily vaccinations in the United Kingdom on 15th March 2021 was: {check2}')

SPOTCHECK: The sum of daily vaccinations in England, Scotland, Wales and Nothern Ireland on 15th March 2021 was: 389942.0
SPOTCHECK: The total daily vaccinations in the United Kingdom on 15th March 2021 was: 389942.0


Since the spot-check showed that United Kingdom duplicates England, Scotland, Wales, Northern Ireland (as expected) rows containing the latter 4 will be dropped to give a total of 147 countries. Since 147 countries are difficult to visualise, the countries will be grouped into regions, keeping those we wish to investigate further separate, e.g. Asia, Europe, but UK & British Overseas, China, Japan, Israel, Russia, USA.

In [27]:
#Delete all rows for England, Scotland, Wales, Northern Ireland
df_processed = df_processed[df_processed.country.apply(lambda x: x not in ['England', 'Scotland', 'Wales', 'Northern Ireland'])]

- [X] Checked 'country' column in detail and removed duplicates (England, United Kingdom, Northern Ireland, Wales, Scotland)

***

**SUMMARY - FEATURE ENGINEERING TASKS** 

- [ ] Add new column for daily vaccinations in %

- [ ] Add population column

- [ ] Add partly vaccinated column

- [ ] Add vaccination type column

- [ ] Consider grouping countries, e.g. by continent

***

***
***
***

TO DO: Dealing with empties

***
***
***

In [28]:
df_processed.isnull().sum(axis = 0)

country                                   0
date                                      0
covaxin                                   0
epivaccorona                              0
johnson_johnson                           0
moderna                                   0
oxford_astrazeneca                        0
pfizer_biontech                           0
sinopharm_beijing                         0
sinopharm_wuhan                           0
sinovac                                   0
sputnik_v                                 0
iso_code                                  0
total_vaccinations                     2996
people_vaccinated                      3557
people_fully_vaccinated                4947
daily_vaccinations                      184
total_vaccinations_per_hundred         2996
people_vaccinated_per_hundred          3557
people_fully_vaccinated_per_hundred    4947
daily_vaccinations_per_million          184
dtype: int64

***

### 2.3 Feature engineering to prepare for in-depth data analysis

In this section we undertake feature engineering to a) map countries to custom regions, b) ...

In [29]:
#Add new column for daily vaccinations in %
df_processed['daily_vaccinations_per_hundred'] = df_processed['daily_vaccinations_per_million']/10000

- [X] Added new column for daily vaccinations in %

In [30]:
#ISSUE - EMPTIES - Merge with kaggle population dataset?
#Generate new column for total population
df_processed['population'] = round(df_processed['people_vaccinated'] * (100 / df_processed['people_vaccinated_per_hundred']), 0)
df_processed['population'] = df_processed['country'].map(df_processed.groupby('country')['population'].mean())

df_processed.population.nunique()

#df_processed.population.isnull().sum()

108

***
***
***

TODO: Finish population column 108 unique value, need 147! check empties

***
***
***

NEW TASK! Add new column reflecting 3 types of vaccines, RNA versus classical versus protein based

In [31]:
#Generate new columns for the three types of vaccines,
df_processed['classical'] = df_processed.covaxin | df_processed.johnson_johnson | df_processed.oxford_astrazeneca | df_processed.sinopharm_beijing | df_processed.sinopharm_wuhan | df_processed.sinovac | df_processed.sputnik_v
df_processed['rna'] = df_processed.pfizer_biontech | df_processed.moderna
df_processed['protein'] = df_processed.epivaccorona
df_processed[['classical', 'rna', 'protein']].sample(5, random_state=42)

classical  rna  protein
6761          1    0        0
3434          1    0        0
5931          0    1        0
6165          1    0        1
1731          0    1        0

#NEW TASK! create new column reflecting partially vaccinated people

In [32]:
#Generate new columns reflecting partially vaccinated people
df_processed['people_partially_vaccinated'] = df_processed['people_vaccinated'] - df_processed['people_fully_vaccinated']
df_processed[['people_partially_vaccinated', 'people_vaccinated', 'people_fully_vaccinated']].sample(5, random_state=42)
#TODO - need to deal with NaNs first!!!!

people_partially_vaccinated  people_vaccinated  people_fully_vaccinated
6761                          NaN                NaN                      NaN
3434                   29646533.0         37125187.0                7478654.0
5931                     184571.0           432414.0                 247843.0
6165                          NaN                NaN                      NaN
1731                          NaN                NaN                      NaN

In [33]:
#Generate new columns reflecting partially vaccinated people
df_processed['people_partially_vaccinated_per_hundred'] = df_processed['people_vaccinated_per_hundred'] - df_processed['people_fully_vaccinated_per_hundred']
df_processed[['people_partially_vaccinated_per_hundred', 'people_vaccinated_per_hundred', 'people_fully_vaccinated_per_hundred']].sample(10, random_state=42)
#TODO - need to deal with NaNs first!!!!

people_partially_vaccinated_per_hundred  people_vaccinated_per_hundred  \
6761                                      NaN                            NaN   
3434                                     2.15                           2.69   
5931                                     1.81                           4.24   
6165                                      NaN                            NaN   
1731                                      NaN                            NaN   
7549                                      NaN                            NaN   
4716                                      NaN                            NaN   
6603                                      NaN                           0.78   
37                                        NaN                            NaN   
4056                                      NaN                            NaN   

      people_fully_vaccinated_per_hundred  
6761                                  NaN  
3434                                 0.54  
5931                                 2.43  
6165                                  NaN  
1731                                  NaN  
7549                                  NaN  
4716                                  NaN  
6603                                  NaN  
37                                    NaN  
4056                                  NaN

In [34]:
#Map countries to custom regions - noteworthy dictionary entries: Europe (EU plus more, munis UK), UK & British Overseas, China, India, Japan, Russia, Israel
#Create lists with countries per region
Africa = ('Algeria','Angola',"Cote d'Ivoire",'Egypt','Equatorial Guinea','Ghana','Guinea','Kenya','Malawi','Maldives','Mali','Mauritius','Morocco',
          'Mozambique','Namibia','Niger','Nigeria','Rwanda','Sao Tome and Principe', 'Senegal','Seychelles','Sierra Leone','Somalia','South Africa',
          'South Sudan','Sudan','Swaziland','Tanzania','Togo','Tunisia','Uganda','Zambia','Zimbabwe')

Asia = ('Afghanistan','Azerbaijan','Bangladesh','Bhutan','Brunei','Cambodia','East Timor','Hong Kong','Indonesia','Kazakhstan','Kyrgyzstan','Laos',
        'Macao','Malaysia','Maldives','Mongolia','Myanmar','Nepal','North Korea','Pakistan','Philippines','Singapore','South Korea','Sri Lanka',
        'Taiwan','Tajikistan','Thailand','Turkmenistan','Uzbekistan','Vietnam')
    
Middle_East = ('Bahrain','Iran','Iraq','Jordan','Kuwait','Lebanon','Oman','Qatar','Saudi Arabia','Syria','Turkey','United Arab Emirates','Yemen')
   
Europe = ('Albania','Andorra','Armenia','Austria','Belarus','Belgium','Bosnia and Herzegovina','Bulgaria','Croatia','Cyprus','Czechia','Denmark','Estonia',
          'Faeroe Islands','Finland','France','Georgia','Germany','Greece','Greenland','Hungary','Iceland','Ireland','Italy','Latvia','Liechtenstein',
          'Lithuania','Luxembourg','Macedonia','Malta','Moldova','Monaco','Montenegro','Netherlands','North Macedonia','Northern Cyprus','Norway','Poland',
          'Portugal','Romania','San Marino','Serbia','Slovakia','Slovenia','Spain','Sweden','Switzerland','Ukraine','Vatican City')

UK_BritishOversea= ('United Kingdom','Anguilla','Antigua and Barbuda','Bermuda','Cayman Islands','Falkland Islands','Gibraltar','Guernsey','Isle of Man',
                    'Jersey', 'Montserrat', 'Saint Helena', 'Turks and Caicos Islands')

North_America = ('Canada','United States')

Central_America_Caribic = ('Antigua and Barbuda','Bahamas','Barbados','Belize','Costa Rica','Cuba','Dominica','Dominican Republic','El Salvador','Grenada',
                           'Guatemala','Haiti','Honduras','Jamaica','Mexico','Nicaragua','Panama','Saint Kitts and Nevis','Saint Lucia',
                           'Saint Vincent and the Grenadines','Trinidad and Tobago')

South_America = ('Argentina','Bolivia','Brazil','Chile','Colombia','Ecuador','Guyana','Paraguay','Peru','Suriname','Uruguay','Venezuela')

Australia_Oceania = ('Australia','Fiji','Kiribati','Marshall Islands','Micronesia','Nauru','New Zealand','Palau','Papua New Guinea','Samoa','Solomon Islands',
                     'Tonga','Tuvalu','Vanuatu')

#Create dictionaries for mapping (incl. specific individual countries)
continents = {'China': 'China', 'India': 'India', 'Israel': 'Israel', 'Japan': 'Japan', 'Russia': 'Russia'}
continents.update({country: 'Asia' for country in Asia})
continents.update({country: 'Europe' for country in Europe})
continents.update({country: 'Africa' for country in Africa})
continents.update({country: 'Middle_East' for country in Middle_East})
continents.update({country: 'UK_BritishOversea' for country in UK_BritishOversea})
continents.update({country: 'North_America' for country in North_America})
continents.update({country: 'Central_America_Caribic' for country in Central_America_Caribic})
continents.update({country: 'South_America' for country in South_America})
continents.update({country: 'Australia_Oceania' for country in Australia_Oceania})

#Create new column and map region to country:
df_processed['country_grouped'] = df_processed['country'].map(continents)

In [35]:
df_processed['country_grouped'].unique()

array(['Asia', 'Europe', 'Africa', 'UK_BritishOversea',
       'Central_America_Caribic', 'South_America', 'Australia_Oceania',
       'Middle_East', 'North_America', 'China', 'India', 'Israel',
       'Japan', 'Russia'], dtype=object)

In [36]:
ncountry = df_processed['country_grouped'].nunique()
print(F'{ncountry} country-groups are included in the revised Covid-19 vaccination progress dataset.')

14 country-groups are included in the revised Covid-19 vaccination progress dataset.


- [X] Mapped countries to custom regions, such as continents, but keeping countries of interest separate, e.g. EU, UK & British Overseas, China, Japan, Israel, Russia, USA

Finally, save the four new cleaned and feature engineered dataframes.

In [37]:
#Save dataframes for use in new notebooks
#Cleaned data
df_processed.to_csv('data/df_processed.csv', index=False)

***

**SUMMARY**

We now have a dataset containing the following:

| FEATURE | description | comment |
|----|----|----|
| **country** | 147 countries | 4 duplicates, UN recognises 197 countries (193 are members) |
| **country_grouped** | 17 countries/groups | NEW column added during feature engineering |
| **date** | data entry (13th December 2020 to 24th March 2021) | - |
| **weekday** | day of week (e.g. Monday) | NEW column added during feature engineering |
| **vaccines** | name/company | - |
| **daily_vaccinations** | no. of vaccination for that date/country | some days only daily vaccinations, some days only the (cummulative) total |
| **daily_vaccinations_per_million** | ratio (in ppm) between vaccination number and total population for the current date in the country | - |
| **daily_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population for the current date in the country | NEW column added during feature engineering |
| **total_vaccinations** | absolute number of immunisations (jabs) in the country | some days only daily vaccinations, some days only the (cummulative) total |
| **total_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population up to the date in the country | - |
| **people_vaccinated** | total no. of people with immunisation (partial and full) | - |
| **people_vaccinated_per_hundred** | ratio (in percent) between population immunized (partial and full) and total population | - |
| **people_fully_vaccinated** | total no. of people with full immunisation (typically 2) | - |
| **people_fully_vaccinated_per_hundred** | ratio (in percent) between population fully immunized and total population | - |
| **people_partially_vaccinated** | total no. of people with partial immunisation (i.e. 1 jab) | - |
| **people_partially_vaccinated_per_hundred** | ratio (in percent) between population partially immunized and total population | - |
| **population** | country/country-group total population | NEW column added during feature engineering |

***
***

# NEW - global dataset

In [38]:
#Reduce dataset to appropriate columns
#(note: % columns ommited as global values cannot be calculated from this dataset)
cols = ['date', 'covaxin', 'epivaccorona', 'johnson_johnson', 'moderna', 'oxford_astrazeneca',
        'pfizer_biontech', 'sinopharm_beijing', 'sinopharm_wuhan', 'sinovac', 'sputnik_v',
        'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'daily_vaccinations',
        'people_partially_vaccinated']
df_global = df_processed[cols]

#Sum up all columns to get global vaccination data
df_global = df_global.groupby('date').sum()

In [39]:
#Save dataframe for combining with Tweets
df_global.to_csv('data/df_global.csv')